File: src/main/java/io/github/slimefunguguproject/bump/implementation/listeners/DragonBreathListener.java
package io.github.slimefunguguproject.bump.implementation.listeners;

import org.bukkit.Particle;
import org.bukkit.entity.AreaEffectCloud;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;

/**
 * This {@link Listener} is responsible for dragon breath damage.
 *
 * @author ybw0014
 */
public class DragonBreathListener implements Listener {
    @EventHandler
    public void onPlayerDamaged(EntityDamageByEntityEvent e) {
        if (e.getDamager() instanceof AreaEffectCloud damager
            && damager.getParticle() == Particle.DRAGON_BREATH
            && damager.getSource() instanceof Player damageSource
            && e.getEntity() instanceof Player p
            && damageSource.getUniqueId().equals(p.getUniqueId())
        ) {
            e.setCancelled(true);
        }
    }
}

File: src/main/kotlin/io/github/slimefunguguproject/bump/implementation/listeners/DragonBreathListener.kt
package io.github.slimefunguguproject.bump.implementation.listeners

import io.github.slimefunguguproject.bump.Bump
import org.bukkit.Particle
import org.bukkit.entity.AreaEffectCloud
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageByEntityEvent

class DragonBreathListener(plugin: Bump) : Listener {
    init {
        plugin.server.pluginManager.registerEvents(this, plugin)
    }

    @EventHandler
    fun onPlayerDamaged(e: EntityDamageByEntityEvent) {
        val damager = e.damager
        if (damager is AreaEffectCloud
            && damager.particle == Particle.DRAGON_BREATH
            && damager.source is Player
            && e.entity is Player
            && (damager.source as Player).uniqueId == e.entity.uniqueId
        ) {
            e.isCancelled = true
        }
    }
}
